p2p/client.go

```diff
@@ -431,6 +534,96 @@ func (c *Client) bootstrapLoop(ctx context.Context) {
 	}
 }
 
+// latest seen height to be retrieved using block-sync
+func (c *Client) setLatestSeenHeight(height uint64) {
+	c.latestSeenHeight = max(height, c.latestSeenHeight)
+}
+
+// retrieveBlockSyncLoop checks if there is any missing block to be retrieved and requests it on demand
+func (c *Client) retrieveBlockSyncLoop(ctx context.Context) {
+	ticker := time.NewTicker(c.conf.BlockSyncRetrieveRetryTime)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			c.blocksyncMu.Lock()
+			state, err := c.store.LoadState()
+			if err != nil {
+				continue
+			}
+
+			// this loop iterates and retrieves all the blocks between the last block applied and the greatest height received,
+			// skipping any block cached, which are the blocks missing to be in sync.
+			for h := state.NextHeight(); h <= c.latestSeenHeight; h++ {
+				_, found := c.heightToSkip[h]
+				if found {
+					continue
+				}
+				c.logger.Debug("Getting block", "height", h)
+				bid, err := c.GetBlockId(ctx, h)
+				if err != nil || bid == cid.Undef {
+					c.logger.Error("unable to find cid", "height", h)
+					continue
+				}
+				_, err = c.store.SaveBlockCid(h, bid, nil)
+				if err != nil {
+					c.logger.Error("unable to store block cid", "height", h, "cid", bid)
+					continue
+				}
+				block, err := c.blocksync.GetBlock(ctx, bid)
+				if err != nil {
+					continue
+				}
+				if c.blocksync.msgHandler != nil {
+					c.blocksync.msgHandler(&block)
+				}
+			}
+			c.blocksyncMu.Unlock()
+		}
+	}
+}
+
+// Blocks content identifiers (CID) are advertised to the DHT by the sequencer before gossiping the block, so a node can find a CID corresponding to a specific height.
+// advertiseBlockSyncLoop re-advertises known CIDs in the DHT. This prevents CIDs are lost in case of nodes disconnections to make sure they can be discovered in the DHT.
+func (c *Client) advertiseBlockSyncLoop(ctx context.Context) {
+	state, err := c.store.LoadState()
+	if err != nil {
+		c.logger.Error("loading state", "err", err)
+		return
+	}
+	ticker := time.NewTicker(c.conf.BlockSyncAdvRetryTime)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+
+			for h := state.BaseHeight; h <= state.Height(); h++ {
+				id, err := c.GetBlockId(ctx, h)
+				if err == nil && id != cid.Undef {
+					continue
+				}
+
+				id, err = c.store.LoadBlockCid(h)
+				if err != nil || id == cid.Undef {
+					continue
+				}
```

Comment:
need concurrency protection on the store?

Reply:
is badgedb not already protecting?

Reply:
Yes, sure 👍